arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -78,251 +126,138 @@ void arm_core_mpu_disable(void)
- */
-static inline void _disable_region(u32_t r_index)
-{
-	/* Attempting to configure MPU_RNR with an invalid
-	 * region number has unpredictable behavior. Therefore
-	 * we add a check before disabling the requested MPU
-	 * region.
-	 */
-	__ASSERT(r_index < _get_num_regions(),
-		"Index 0x%x out-of-bound (supported regions: 0x%x)\n",
-		r_index,
-		_get_num_regions());
-	LOG_DBG("disable region 0x%x", r_index);
-	/* Disable region */
-	ARM_MPU_ClrRegion(r_index);
+		/* Region found */
+		reg_index = i;
+		break;
+	}
+	__ASSERT(reg_index != _get_num_regions(),
+		"Memory domain partition not found\n");
+
+	/* Modify the permissions */
+	partition->attr = *new_attr;
+	_mpu_configure_region(reg_index, partition);
 }
 
 /**
- * @brief configure the base address and size for an MPU region
- *
- * @param   type    MPU region type
- * @param   base    base address in RAM
- * @param   size    size of the region
+ * @brief get the maximum number of free regions for memory domain partitions
  */
-void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
+int arm_core_mpu_get_max_domain_partition_regions(void)
 {
-	struct arm_mpu_region region_conf;
-
-	LOG_DBG("Region info: 0x%x 0x%x", base, size);
-	u32_t region_index = _get_region_index_by_type(type);
-
-	if (_get_region_attr_by_type(&region_conf.attr, type, base, size)) {
-		return;
-	}
-	region_conf.base = base;
-
-	if (region_index >= _get_num_regions()) {
-		return;
-	}
-
-	_region_init(region_index, &region_conf);
-}
+	int available_regions_num =	_get_num_regions() - static_regions_num;
 
-#if defined(CONFIG_USERSPACE)
-void arm_core_mpu_configure_user_context(struct k_thread *thread)
-{
-	if (!thread->arch.priv_stack_start) {
-		_disable_region(_get_region_index_by_type(
-			THREAD_STACK_REGION));
-		return;
-	}
+	/* Additional regions required for the thread stack,
+	 * depending on whether the MPU architecture requires
+	 * non overlapping regions.
+	 */
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
+	available_regions_num -= 2;
+#else
+	available_regions_num -= 1;
+#endif
 
-	u32_t base = (u32_t)thread->stack_obj;
-	u32_t size = thread->stack_info.size;
-#if !defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
-	/* In user-mode the thread stack will include the (optional)
-	 * guard area. For MPUs with arbitrary base address and limit
-	 * it is essential to include this size increase, to avoid
-	 * MPU faults.
+#if defined(CONFIG_MPU_STACK_GUARD)
+	/* Additional regions required for the current thread's privileged
+	 * stack guard, depending on whether the MPU architecture requires
+	 * non overlapping regions.
 	 */
-	size += thread->stack_info.start - base;
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
+	available_regions_num -= 2;
+#else
+	available_regions_num -= 1;
+#endif
 #endif
 
-	arm_core_mpu_configure(THREAD_STACK_REGION, base, size);
+	return _MPU_MAX_DOMAIN_PARTITIONS_GET(available_regions_num);
```

Comment:
So where the number of regions should be adjusted to take under account HW and SW config? Above or in this macro?

Reply:
In this macro. Note that the macro is defined for all the three different driver implementations.

Reply:
(addressed)